src/site/content/en/reliable/measuring-offline-usage/index.md

```diff
@@ -0,0 +1,227 @@
+[many browsers support](https://caniuse.com/#feat=online-status)) and to put your analytics tracking
+logic in those listeners. Unfortunately, there are several problems and limitations with this
+approach:
+
++   In general tracking every network connection status event might be excessive, and is
+    counter-productive in a privacy-centric world where as little data as possible should be
+    collected. Additionally the `online` and `offline` events can fire for just a split second of
+    network loss, which a user probably wouldn't even see or notice.
++   The analytics tracking of offline activity would never reach the analytics server because
+    the user is… well, offline.
++   Tracking a timestamp locally when a user goes offline and sending the offline activity to
+    the analytics server when the user goes back online depends on the user revisiting your site.
+    If the user drops off your site due to a lack of an offline mode and never revisits, you have
+    no way to track that. The ability to track offline drop-offs is critical data for building a
+    case about why your site needs a better offline mode.
++   The `online` event is not very reliable as it
+    [only knows about network access](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorOnLine/onLine),
+    not internet access. Therefore a user might still be offline, and sending the tracking ping can
+    still fail.
++   Even if the user still stays on the current page while being offline, none of the other
+    analytics events (e.g. scroll events, clicks, etc.) are tracked either, which might be the more
+    relevant and useful information.
++   Being offline in itself is also not too meaningful in general. As a website developer it may
+    be more important to know what kinds of resources failed to load. This is especially relevant
+    in the context of SPAs, where a dropped network connection might not lead to a browser offline
+    error page (which users understand) but more likely to random dynamic parts of the page failing
+    silently.
+
+You can still use this solution to gain a basic understanding of offline usage, but the many
+drawbacks and limitations need to be considered carefully.
+
+## A better approach: the service worker
+
+The solution that enables offline mode turns out to be the better solution for tracking offline
+usage.  The basic idea is to store analytics pings into IndexedDB as long as the user is offline,
+and just resend them when the user goes online again. For Google Analytics this is already available
+[off-the-shelf through a Workbox module](https://developers.google.com/web/tools/workbox/modules/workbox-google-analytics),
+but keep in mind that hits sent more than
+[four hours deferred](https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters#qt)
+may not be processed. In its simplest form, it can be activated within a Workbox-based service
+worker with these two lines:
+
+```js
```

Comment:
We support code highlighting if you want to highlight lines in any of these samples: https://web.dev/handbook/markup-code/#code-highlighting